### PR TITLE
Treat Void formal arguments as @Nullable

### PR DESCRIPTION
**Pull request by lazaroclapp**
This is a follow up on #595. See that PR for some of the soundness
caveats related to method overriding and the nullability of Void
in our current implementation.

This does, however, avoid some otherwise awkward suppressions,
as shown in #608.

There are two correct ways of handling Void:

a) Having it default to `@Nullable` as a type, which requires us
   adding a "default nullability" for some types/classes and explicitly
   contradicts JSpecify (jspecify/jspecify#51)
b) Supporting generics, and requiring explicitly annotating `@Nullable Void`.
   This will probably also require third-party libraries which we consider
   annotated (such as CF), to adopt this convention.

I believe (b) is the way forward long-term, which means that, for now,
this hack might be the best we can do without generics support. Once
NullAway supports generics, both this change and #595 should be reverted.

### CODE REVIEW

File: nullaway/src/test/java/com/uber/nullaway/NullAwayCoreTests.java [hunk 1]
```diff
@@ -623,6 +623,29 @@ public void nullableOnJavaLangVoid() {
...
+            "class Test {",
+            "  Void foo1() {",
+            "    // temporarily, we treat a Void return type as if it was @Nullable Void",
+            "    return (Void)null;",
+            "  }",
+            "  void consumeVoid(Void v) {",
+            "  }",
+            "  @Nullable Void foo2() {",
+            "    // temporarily, we treat a Void argument type as if it was @Nullable Void",
+            "    consumeVoid((Void)null);",
```
Comment (msridhar) [inline nullaway/src/test/java/com/uber/nullaway/NullAwayCoreTests.java:642]: From the test, it seems that this change makes us treat `(Void) null` as `@NonNull Void` when it is a parameter, not `@Nullable Void` (since here `consumeVoid`'s argument is `@NonNull`).  Am I missing something?

**Review by msridhar — COMMENTED** [review]
Beyond the comment below, could we add a test for assigning `(Void) null` to a field?  I'm not sure what to expect there (though it is a more contrived case).

File: nullaway/src/test/java/com/uber/nullaway/NullAwayCoreTests.java (hunk 1 above)
Comment (lazaroclapp) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayCoreTests.java:642]: What this is meant to change is the interpretation of a `Void` **formal** to automatically mean `@Nullable Void`. See the change to `nonNullPositions` [here](https://github.com/uber/NullAway/pull/613/files#diff-a53186e00f6c045ebdbe4813d3e1ff0a469095e5034d5273571c09334a6cb4e4R1450).

It is not possible at this time to always annotate the formal as `@Nullable`, because, usually, in the cases where we want to pass `Void` around, the formal is of some generic type `T` and what we really would want is to be able to replace `<Void>` with `<@Nullable Void>`, see the example in https://github.com/uber/NullAway/pull/608.

**Comment by lazaroclapp** [comment]
> Beyond the comment below, could we add a test for assigning `(Void) null` to a field? I'm not sure what to expect there (though it is a more contrived case).

Will try, though I'd expect that to fail within this hack. If we want a more robust fix here, we might need some sort of mechanism to specify the default nullness of a type. Basically, forcing NullAway to expand `Void` into `@Nullable Void` everywhere (for now), then revert that once we have support for annotations on generics.

File: nullaway/src/test/java/com/uber/nullaway/NullAwayCoreTests.java (hunk 1 above)
Comment (msridhar) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayCoreTests.java:642]: Got it thanks! I was thinking about the actual rather than the formal

File: nullaway/src/test/java/com/uber/nullaway/NullAwayCoreTests.java [hunk 2]
```diff
@@ -623,6 +623,29 @@ public void nullableOnJavaLangVoid() {
...
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "class Test {",
+            "  Void foo1() {",
+            "    // temporarily, we treat a Void return type as if it was @Nullable Void",
+            "    return (Void)null;",
+            "  }",
+            "  void consumeVoid(Void v) {",
```
Comment (msridhar) [inline nullaway/src/test/java/com/uber/nullaway/NullAwayCoreTests.java:638]: Can we add a comment here indicating the formal is treated as if it were `@Nullable`? I think having the comment just at the call site confused me a bit

File: nullaway/src/test/java/com/uber/nullaway/NullAwayCoreTests.java (hunk 1 above)
Comment (msridhar) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayCoreTests.java:642]: Just doing `consumeVoid(null)` should also have no error right? Can we test that?

**Review by msridhar — APPROVED** [review]
Code change looks good, just a couple comments on the test
